### DISCUSSION TITLE
automatic database conversion on upgrade

### DISCUSSION THREAD

**Discussion by arekm** [sysstat/sysstat#298] [Q&A]
Hello.

How packages, when upgraded, should deal with incompatible database format conversion? (in automated way)

I was for example doing something like

````
for log in /var/log/sa/sa[0-9]*; do
        if (LC_ALL=C %{_bindir}/sadf -C "$log" 2>&1 | grep -q "Current sysstat version cannot read the format of this file"); then
                echo "Converting file $log to current format: "
                if (%{_bindir}/sadf -c "$log" > "$log.migrate"); then
                        chown --reference "$log" "$log.migrate"
                        chmod --reference "$log" "$log.migrate"
                        mv "$log.migrate" "$log"
                else
                        echo "$log MIGRATION FAILED." >&2
                fi
        fi
done
````

but it's not a good method since

````
LC_ALL=C /usr/bin/sadf -C "$log"
````
no longer complains (sysstat 12.2.1) while sadc complains

````
# LC_ALL=C /usr/lib/sa/sadc -
Invalid system activity file: /var/log/sa/sa28
File created by sar/sadc from sysstat version 10.0.2
Current sysstat version cannot read the format of this file (0x2175)
````

Is there a better way?

Edit:
Tested 12.5.4 with the same result. sadc complains, sadf doesn't.

Some "sadf --migrate-database" that would find all files and just did migration would be nice (command run unconditionally on each sysstat upgrade, so no need to figure out if upgrade is needed or not)

**Comment by sysstat** [comment]
Hi,
I don't understand: `sadf` actually complains when asked to read a file from an old sysstat version:

```
$ ./sadf -V
sysstat version 12.2.1
(C) Sebastien Godard (sysstat <at> orange.fr)
$ ./sadf -C data-10.0.2
Invalid system activity file: data-10.0.2
File created by sar/sadc from sysstat version 10.0.2
Current sysstat version cannot read the format of this file (0x2171)
```
```
$ ./sadf -V
sysstat version 12.5.4
(C) Sebastien Godard (sysstat <at> orange.fr)
$ ./sadf -C data-10.0.2
Invalid system activity file: data-10.0.2
File created by sar/sadc from sysstat version 10.0.2
Current sysstat version cannot read the format of this file (0x2171)
```

**Reply by arekm** [reply]
Not here

````
# LC_ALL=C /usr/bin/sadf -C /var/log/sa/sa28
box1	-1	2012-01-28 04:05:57 UTC	LINUX-RESTART	(2 CPU)
box1	-1	2012-01-28 04:10:37 UTC	LINUX-RESTART	(2 CPU)
box1	-1	2012-01-28 04:33:08 UTC	LINUX-RESTART	(2 CPU)
# LC_ALL=C /usr/lib/sa/sadc -
Invalid system activity file: /var/log/sa/sa28
File created by sar/sadc from sysstat version 10.0.2
Current sysstat version cannot read the format of this file (0x2175)
# LC_ALL=C /usr/bin/sadf -V
sysstat version 12.5.4
(C) Sebastien Godard (sysstat <at> orange.fr)
````

**Reply by sysstat** [reply]
Your datafile has been converted to version 12.x format. I can say this because the FORMAT_MAGIC number of your sa28 datafile is **0x2175**, whereas it is **0x2171** for a genuine file created by sysstat 10.0.2.
That's why `sadf` doesn't complain: Once converted, the file can be read by `sar` and `sadf`.
Wrt to `sadc`, it will never append data to a file which has not been created by the same sysstat's version as his. It complains if told to do so.

Note: `sadf -H` will tell you which version of sysstat has been used to create a file:
```
$ sadf -V
sysstat version 12.5.4
(C) Sebastien Godard (sysstat <at> orange.fr)
$ sadf -H data-10.0.2
System activity data file: data-10.0.2 (0x2171)
File created by sar/sadc from sysstat version 10.0.2
$ sadf -H /var/log/sa/sa28 | grep "File created"
File created by sar/sadc from sysstat version 12.5.4
```

**Reply by arekm** [reply]
Ok, so I should just always try to convert I guess, on each upgrade (even with minor version or from the same version to the same version for sake of simplicity) as solution.

Not so easy. In case if saXX file is in current format:
````
sadf -c log > newlog
````
will succeed (exit status 0!) and will create 0 sized output file (+ message that file is up to date). thus mv newlog log will destroy data.

So exit status of sadf cannot be used reliably, too.

Edit1:

Current script:
````
# Always try conversion; format changes unreliably and there is no easy
# and automated way to find out when conversion is needed
for log in /var/log/sa/sa[0-9]*; do
        if (%{_bindir}/sadf -c "$log" > "$log.migrate"); then
                if [ -s "$log.migrate" ]; then
                        # assume format upgrade was done correctly
                        chown --reference "$log" "$log.migrate"
                        chmod --reference "$log" "$log.migrate"
                        mv "$log.migrate" "$log"
                else
                        # assume format is already up to date
                        rm -f "$log.migrate"
                fi
        else
                # conversion failed, so this is useless file
                rm -f "$log.migrate"
                echo "Converting file $log to current format: FAILED." >&2
        fi
done
`````

Still not good enough because will spam stderr with "File format already up-to-date" (a message that we don't care) but we can't redirect it to /dev/null because we care about other errors.

Edit2:

ehm, not good at all

````
# /usr/lib/sa/sadc -
Invalid system activity file: /var/log/sa/sa29
File created by sar/sadc from sysstat version 10.0.2
Current sysstat version cannot read the format of this file (0x2175)
# sadf -c /var/log/sa/sa29 > x

File format already up-to-date
````

Back to my question. How to get format upgrade in automated way reliably, so I can query old files and sadc to be able to store new data?
